### PR TITLE
[FIX] web_editor: apply format with tag

### PR DESCRIPTION
**Pull request by Goaman**
Before this commit
The editor used only span with a style tag in order to format text.
This was considered to be unclean code.

After this commit
- use style tag (strong, em, s, u) rather than span with a style tag
- use span with style tag if a container is already styled
- remove redundant style/tags

task-2850566

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/web_editor/static/lib/odoo-editor/src/utils/utils.js [hunk 1]
```diff
@@ -755,6 +755,252 @@ export function preserveCursor(document) {
...
+    switchDirection: {
+        isFormatted: isDirectionSwitched,
+    }
+
+}
+
+const isBlockOrHasClass = (node) => {
+    return isBlock(node) || (node.classList && Boolean(node.classList.length));
+}
+const ancestorsUntilBlockOrHasClass = (node) => {
```
Comment (dmo-odoo) [inline addons/web_editor/static/lib/odoo-editor/src/utils/utils.js:814]: This is used only once outside of its own recursivity. Couldn't it be replaced by a loop at the call site ?

File: addons/web_editor/static/lib/odoo-editor/src/utils/utils.js [hunk 2]
```diff
@@ -755,6 +755,252 @@ export function preserveCursor(document) {
...
+    if (node.getAttribute('style') === '') {
+        node.removeAttribute('style');
+    }
+};
+const removeStyleItem = (node, styleName, item) => {
+    const newStyle = node.style[styleName].split(' ').filter(x => x !== item).join(' ');
+    node.style[styleName] = newStyle || null;
+    if (node.getAttribute('style') === '') {
+        node.removeAttribute('style');
+    }
```
Comment (dmo-odoo) [inline addons/web_editor/static/lib/odoo-editor/src/utils/utils.js:829]: I suggest to merge this function into `removeStyle`. If an item is provided, it only removes the provided item. Otherwise, it removes the entire style.

File: addons/web_editor/static/lib/odoo-editor/src/utils/utils.js [hunk 3]
```diff
@@ -1429,29 +1682,34 @@ export function splitElement(element, offset) {
...
         beforeSplit = splitElement(before.parentElement, childNodeIndex(before) + 1)[1];
-        newUntil = beforeSplit;
-        before = newUntil.previousSibling;
+        before = beforeSplit.previousSibling;
     }
     if (before) {
         beforeSplit = splitElement(limitAncestor, childNodeIndex(before) + 1)[1];
     }
-    return [beforeSplit, afterSplit];
+    return beforeSplit || afterSplit || limitAncestor;
```
Comment (dmo-odoo) [inline addons/web_editor/static/lib/odoo-editor/src/utils/utils.js:1712]: That is quite the change in the util here. I'm curious about it so I'd be interested in discussing this together tomorrow.

File: addons/web_editor/static/lib/odoo-editor/test/spec/editor.test.js [hunk 4]
```diff
@@ -3371,12 +3370,8 @@ X[]
                     stepFunction: async editor => {
                         // simulate preview
                         editor.historyPauseSteps();
-                        try {
-                            applyInlineStyle(editor, (el) => el.style.color = 'lime');
-                            // a[bcd]e with bcd in lime
-                        } finally {
-                            editor.historyUnpauseSteps();
-                        }
+                        editor.execCommand('bold');
```
Comment (dmo-odoo) [inline addons/web_editor/static/lib/odoo-editor/test/spec/editor.test.js:3373]: Why the change ?

File: addons/web_editor/static/lib/odoo-editor/test/spec/format.test.js [hunk 5]
```diff
@@ -13,177 +12,275 @@ const underline = async editor => {
...
+    const getZwsTag = (tagName, {style} = {}) => {
+        const styleAttr = style ? ` style="${style}"` : '';
+        return (content, zws) => {
+            const zwsFirstAttr = zws === 'first' ? ' oe-zws-empty-inline=""' : '';
+            const zwsLastAttr = zws === 'last' ? ' oe-zws-empty-inline=""' : '';
+            return `<${tagName}${zwsFirstAttr}${styleAttr}${zwsLastAttr}>${content}</${tagName}>`
+        };
+    }
+
+    const execWith = (functions) => {
```
Comment (dmo-odoo) [inline addons/web_editor/static/lib/odoo-editor/test/spec/format.test.js:27]: This is used only once.

File: addons/web_editor/static/lib/odoo-editor/test/spec/format.test.js [hunk 6]
```diff
@@ -13,177 +12,275 @@ const underline = async editor => {
...
+            }
+        }
+    }
+
+    const span = getZwsTag('span');
+
+    const strong = getZwsTag('strong');
+    const notStrong = getZwsTag('span', {style: 'font-weight: normal;'});
+    const spanBold = getZwsTag('span', {style: 'font-weight: bolder;'});
+    const b = (content) => `<b>${content}</b>`;
```
Comment (dmo-odoo) [inline addons/web_editor/static/lib/odoo-editor/test/spec/format.test.js:40]: I'm confused. Why is `b` treated differently than `strong` ? One gets a special "zws" treatment but not the other.

File: addons/web_editor/static/lib/odoo-editor/test/spec/format.test.js [hunk 7]
```diff
@@ -13,177 +12,275 @@ const underline = async editor => {
...
+    const s = getZwsTag('s');
+
+
     describe('bold', () => {
         it('should make a few characters bold', async () => {
             await testEditor(BasicEditor, {
                 contentBefore: '<p>ab[cde]fg</p>',
                 stepFunction: bold,
-                contentAfter: `<p>ab${b(`[cde]`)}fg</p>`,
+                contentAfter: `<p>ab${strong(`[cde]`)}fg</p>`,
```
Comment (dmo-odoo) [inline addons/web_editor/static/lib/odoo-editor/test/spec/format.test.js:55]: We should discuss tomorrow with the whole team if we prefer to have `b` or `strong` as the default as Fabien didn't mind one or the other.

File: addons/web_editor/static/lib/odoo-editor/test/spec/format.test.js [hunk 8]
```diff
@@ -398,41 +494,42 @@ describe('Format', () => {
...
+                contentAfter: `<p>ab${u(s(`cd[]ef`))}</p>`,
             });
         });
         it('should remove underline after restoring it after removing it (collapsed, strikeThrough)', async () => {
             await testEditor(BasicEditor, {
                 contentBefore: `<p>ab${u(s(`cd`))}${s(u(`[]\u200b`, 'first'))}${u(s(`ef`))}</p>`,
                 stepFunction: underline,
-                contentAfterEdit: `<p>ab${u(s(`cd`))}${s(`\u200b[]`, 'last')}${u(s(`ef`))}</p>`,
-                contentAfter: `<p>ab${u(s(`cd`))}[]${u(s(`ef`))}</p>`,
+                contentAfterEdit: `<p>ab${u(s(`cd`))}${s(`[]\u200b`, 'last')}${u(s(`ef`))}</p>`,
```
Comment (dmo-odoo) [inline addons/web_editor/static/lib/odoo-editor/test/spec/format.test.js:524]: Wait, the zws changed place ?

File: addons/web_editor/static/lib/odoo-editor/test/spec/format.test.js [hunk 9]
```diff
@@ -13,177 +12,275 @@ const underline = async editor => {
...
+const setFontSize = size => editor => editor.execCommand('setFontSize', size);
 
 describe('Format', () => {
-    const b = (content, zws) => `<span${zws === 'first' ? ' oe-zws-empty-inline=""' : ''} style="font-weight: bolder;"${zws === 'last' ? ' oe-zws-empty-inline=""' : ''}>${content}</span>`;
-    const notB = (content, weight, zws) => `<span${zws === 'first' ? ' oe-zws-empty-inline=""' : ''} style="font-weight: ${weight || 'normal'};"${zws === 'last' ? ' oe-zws-empty-inline=""' : ''}>${content}</span>`;
+    const getZwsTag = (tagName, {style} = {}) => {
+        const styleAttr = style ? ` style="${style}"` : '';
+        return (content, zws) => {
+            const zwsFirstAttr = zws === 'first' ? ' oe-zws-empty-inline=""' : '';
+            const zwsLastAttr = zws === 'last' ? ' oe-zws-empty-inline=""' : '';
```
Comment (dmo-odoo) [inline addons/web_editor/static/lib/odoo-editor/test/spec/format.test.js:22]: I'm very confused about these "first" and "last" parameters. I know they were there before, and I guess I just glossed over those test utils in a previous review rush, but I'd like to see if we could challenge them together.

File: addons/web_editor/static/lib/odoo-editor/test/utils.js [hunk 10]
```diff
@@ -274,6 +274,12 @@ export async function testEditor(Editor = OdooEditor, spec, options = {}) {
     const testNode = document.createElement('div');
     document.querySelector('#editor-test-container').innerHTML = '';
     document.querySelector('#editor-test-container').appendChild(testNode);
+    let styleTag;
+    if (spec.styleContent) {
```
Comment (dmo-odoo) [inline addons/web_editor/static/lib/odoo-editor/test/utils.js:278]: I like this feature. I'd call it `style` though, like the `<style>` tag that it is actually emulating.

File: addons/web_editor/static/lib/odoo-editor/src/utils/utils.js [hunk 11]
```diff
@@ -755,6 +755,252 @@ export function preserveCursor(document) {
...
+    const newNode = document.createElement(tagName);
+    while (node.firstChild) {
+        newNode.appendChild(node.firstChild); // *Moves* the child
+    }
+    for (let index = node.attributes.length - 1; index >= 0; --index) {
+        newNode.attributes.setNamedItem(node.attributes[index].cloneNode());
+    }
+    node.parentNode.replaceChild(newNode, node);
+}
+const removeFormat = (node, formatInfo) => {
```
[the remark this answers is not in the document]
Comment (Goaman) [reply addons/web_editor/static/lib/odoo-editor/src/utils/utils.js:857]: This method remove a specific format where the editor command remove all the formats through the native browser `execCommand('removeFormat')`.

File: addons/web_editor/static/lib/odoo-editor/src/utils/utils.js (hunk 1 above)
Comment (Goaman) [reply addons/web_editor/static/lib/odoo-editor/src/utils/utils.js:814]: I find the loop more cubersome and more difficult to understand:
```js
        const ancestorsUntilBlockOrHasClass = [];
        let node = selectedTextNode.parentElement;
        while (node && !isBlock(node) && !(node.classList && Boolean(node.classList.length))) {
            ancestorsUntilBlockOrHasClass.push(node);
            node = node.parentElement;
        }
        ancestorsUntilBlockOrHasClass.push(node);
        const [blockOrInlineClass, ...inlineAncestors] = ancestorsUntilBlockOrHasClass.reverse();
```

File: addons/web_editor/static/lib/odoo-editor/test/spec/editor.test.js (hunk 4 above)
Comment (Goaman) [reply addons/web_editor/static/lib/odoo-editor/test/spec/editor.test.js:3373]: Because `applyInlineStyle` is removed.

File: addons/web_editor/static/lib/odoo-editor/test/spec/format.test.js (hunk 6 above)
Comment (Goaman) [reply addons/web_editor/static/lib/odoo-editor/test/spec/format.test.js:40]: It was a mistake. Good catch.

File: addons/web_editor/static/lib/odoo-editor/test/spec/format.test.js (hunk 8 above)
Comment (Goaman) [reply addons/web_editor/static/lib/odoo-editor/test/spec/format.test.js:524]: The selection is slightly different. But it does not change the behavior.

File: addons/web_editor/static/lib/odoo-editor/src/utils/utils.js (hunk 1 above)
Comment (dmo-odoo) [reply addons/web_editor/static/lib/odoo-editor/src/utils/utils.js:814]: You're saying that this :
```js
const isBlockOrHasClass = (node) => {
    return isBlock(node) || (node.classList && Boolean(node.classList.length));
}
const ancestorsUntilBlockOrHasClass = (node) => {
    if (!node || !node.parentElement || isBlockOrHasClass(node)) return [];
    return [node.parentElement, ...ancestorsUntilBlockOrHasClass(node.parentElement)];
};
const [blockOrInlineClass, ...inlineAncestors] = ancestorsUntilBlockOrHasClass(selectedTextNode).reverse();
```
Is better than this :
```js
// Compute inline ancestors until finding one which is a block or has a class.
const ancestors = [];
let node = selectedTextNode.parentElement;
while (node && !isBlock(node) && !(node.classList && node.classList.length)) { 
    ancestors.push(node);
    node = node.parentElement;
} 
ancestors.push(node);
const [blockOrInlineClass, ...inlineAncestors] = ancestors.reverse();
```
I already disagree at that point but, even if I did agree, I would argue that it's because you didn't push the conversion from one version to the other far enough. I'd write it like this:
```js
// Compute inline ancestors until finding one which is a block or has a class.
let blockOrInlineClass;
const inlineAncestors = [];
let node = selectedTextNode;
while (!blockOrInlineClass && (node = node.parentElement)) {
    if (isBlock(node) || (node.classList && node.classList.length)) {
        blockOrInlineClass = node;
    } else {
        inlineAncestors.unshift(node);
    }
} 
```
This code says _exactly_ what it does at each point, does not write `node.parentElement` multiple times, never returns deliberately empty values, does not construct the array in the wrong order just to call reverse at the end, and does not return a tuple. (Because yes, pushing the blockOrInlineClass ancestor in the same array as the other ancestors is somewhat hiding a tuple return type, and the proof is that you have to deconstruct the return value to separate the two different kinds of result returned by the function.)

I guess you'll disagree that it is better than the version with the ultra ad-hoc helpers so what do we do now ? ^^'

File: addons/web_editor/static/lib/odoo-editor/src/utils/utils.js [hunk 12]
```diff
@@ -755,6 +755,253 @@ export function preserveCursor(document) {
...
+        let lastAncestorInlineFormat;
+
+        for (const ancestor of inlineAncestors) {
+            const ancestorIsTag = formatInfo.isTag && formatInfo.isTag(ancestor);
+            const ancestorHasStyle = formatInfo.hasStyle && formatInfo.hasStyle(ancestor, formatProps);
+            const ancestorTagIsFormated = ancestorIsTag || ancestorHasStyle;
+            const isUselessZwsSpan =
+                ancestor.tagName === 'SPAN' &&
+                ancestor.hasAttribute('oe-zws-empty-inline') &&
+                ancestor.getAttributeNames().length === 1;
```
[the remark this answers is not in the document]
Comment (Goaman) [reply addons/web_editor/static/lib/odoo-editor/src/utils/utils.js:919]: Otherwise, a tag could go inside that technical span and remain `forever`.
Example of a test that break without the removal of the useless span:
```
actual  : '<p>ab<span oe-zws-empty-inline=""><em oe-zws-empty-inline="">[]//zws//</em></span>cd</p>'
expected: '<p>ab<em oe-zws-empty-inline="">[]//zws//</em>cd</p>'
```

File: addons/web_editor/static/lib/odoo-editor/src/utils/utils.js (hunk 1 above)
Comment (Goaman) [reply addons/web_editor/static/lib/odoo-editor/src/utils/utils.js:836]: Suggestion applied.

File: addons/web_editor/static/lib/odoo-editor/test/spec/format.test.js (hunk 5 above)
Comment (Goaman) [reply addons/web_editor/static/lib/odoo-editor/test/spec/format.test.js:27]: Yes, I forgot to remove it.

File: addons/web_editor/static/lib/odoo-editor/src/utils/utils.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (Goaman) [reply addons/web_editor/static/lib/odoo-editor/src/utils/utils.js:940]: unwrap zws and remove zws

**Comment by dmo-odoo** [comment]
@robodoo rebase-ff r+
